### ISSUE TITLE
Sealed sender not working

### ISSUE THREAD

**Issue by pgerber** [signalapp/Signal-Android#8339]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
I enabled "Display indicators". Yet, when communicating between Signal Android (4.30.2, linked to three Signal Desktop v1.18.0-beta.5) and Signal Android (4.30.2, no linked devices), I do not see an indicator in the "Message details" view.

### Steps to reproduce
- Send messages forth and back
- Check message details for sealed sender indicator

**Actual result:** No sealed sender indicator
**Expected result:** Sealed sender indicator

#### Side A (with 3 Signal Desktop)

Signal Android (A1):
**Device:** SGP Blackphone 2 (BP2)
**Android version:** Silent OS 3.0.23 / Android 6.0.1
**Signal version:** 4.30.2

Signal Desktop (A2):
**OS:** Debian 9
**Signal version:** v1.18.0-beta.5

Signal Desktop (A3):
**OS:** Debian 9
**Signal version:** v1.18.0-beta.5

Signal Desktop (A4):
**OS:** Debian 9 template VM on Qubes OS 4
**Signal version:** v1.18.0-beta.5

#### Side B (no Signal Desktop)

Signal Android (B1):
**Device:** Samsung Galaxy S5
**Android version:** Lineage 15.1 / Android 8.1.0
**Signal version:** 4.30.2

### Link to debug log
The message were sent between +[REDACTED]275 (A) and +[REDACTED]453 (B)

[A1](https://debuglogs.org/a19d0a1952e60bb6ba80ca0e92e457fa0134df65a84fa3e35200b54b5f830fc2)
[A2](https://debuglogs.org/b80326e6f2ca4efd2fdf423fb4a6dbf7b6840bd9d20df43f1e3ffe6808921ed2)
[A3](https://debuglogs.org/6cf28bf2b919c45c5a99b95e345e6a08c27996a5688d47a1fea5de84992d4f08)
~~A4~~ (I don't have logs ATM, but I can get them if need be.)
[B1](https://debuglogs.org/5b0d3b2b87f0f42c3c7db8adc6ed6a13a8c13333be2df673fa9d31ffa7036792)

**Comment by jpds** [comment]
I've also noticed this with my chats - perhaps related to e31029da89059f1ea21ad6b45aba1f76c94c12aa ?

**Comment by minnmann** [comment]
I have similar behavior for one contact. As none of my other contacts use the beta release, I can't tell if it's only for one contact or a general issue.

Two differences in my case:
- Sealed sender messages used to work from the initial release of the beta until a few days ago.
- Incoming messages from the contact are still shown as sealed sender messages.

If I should post as a separate issue, please let me know.

Some more information/observations:
- The contact updated the profile on 10/31/2018. The sealed sender messages still worked after that.
- I did an explicit contact refresh on my phone without any change.
- "Sealed sender indicator" switched off and on without any change.
- "Allow from anyone" enabled.

Debug log:
https://debuglogs.org/a61ab94734e6fa4a77363fa76b5e38ce58d039d1263acddd8ea7c0f934d3cc58

The contact in question is +***********73
My user should be +**********66.

Last outgoing sealed sender message: 11/02/2018, 15:58:28 GMT-07:00
First outgoing message sent as regular message: 11/03/2018 18:08:11 GMT-07:00

As the contact is currently on travel, unfortunately I can not check, if my outgoing messages show as sealed sender messages on the receiving side.
However, on my Signal Desktop instance the messages show the same behavior (incoming sealed sender, outgoing regular), so I'd guess it's not only a visual issue.

**Comment by greyson-signal** [comment]
@pgerber First read of the Android logs is that we're detecting that not all of your linked devices support sealed sender, which causes you not to send sealed sender messages.

I'm going to have desktop look at the logs and see if they notice anything. In the meantime, can you double-check that you don't have *any* other linked devices, and that all of the desktop versions you reported are correct? I super-appreciate that you were so detailed with listing out your linked devices -- this is more just another sanity check :) Thanks!

**Comment by pgerber** [comment]
> can you double-check that you don't have any other linked devices

Done, Android app shows three linked devices for A and none for B, just as expected.

> and that all of the desktop versions you reported are correct?

Checked, all Desktop installations are running 1.18.0-beta.5.

**Comment by greyson-signal** [comment]
@minnmann From your log, it appears you're able to send sealed sender messages just fine. The issue is that the specific person you're contacting does not support sealed sender. This could be because you either don't have an up-to-date profile, or they added a linked desktop instance that doesn't support sealed sender. I can't debug your specific case further until I have logs from the *73 device.

**Comment by minnmann** [comment]
@greyson-signal Thanks a lot! It actually turned out that this contact updated one of her desktop instances exactly in that time frame, but did not wait until all messages have been loaded, probably omitting the config sync messages, if I understand correctly, pushing it back to the "fallback" (old) state.
Now, after loading all messages, it works again.

Just being curious: Even if this person wouldn't support sealed sender on some of her devices, is it correct that all messages sent from her phone are still sealed sender messages, while my outgoing ones are not?
And secondly, there is still at least one desktop instance on her side, which has not yet been updated (offline for several weeks or months). Will she be able to receive my outgoing (sealed sender) messages?

I also got a debug log from her phone. Just let me know, if you still need it.

**Comment by greyson-signal** [comment]
@minnmann An individual should only be able to send and receive sealed sender messages if all of their devices support it. Otherwise, when you sent a sealed sender message to them, their desktop (which doesn't support sealed sender) wouldn't be able to read it. Send over the debug log when you get the chance. Thanks!

**Comment by minnmann** [comment]
@greyson-signal Yeah, that's what I would have imagined, too. The debug log from her phone is here: 
https://debuglogs.org/3957162d7f12f59cad033283ae8827647a178ce1eaf4f14969921492eeb102e2
The 2nd desktop instance has not been online since about mid September according to the linked device overview on Android.

**Comment by greyson-signal** [comment]
@minnmann Got confirmation that when deciding sealed sender eligibility, we exclude linked devices that haven't been used for more than a month. This prevents "forgotten" linked desktops from preventing the sending of sealed sender messages. So sounds like everything is good with your friend. Thanks!

**Comment by greyson-signal** [comment]
@jpds 
> I've also noticed this with my chats - perhaps related to e31029d ?

Unlikely, given that I haven't seen that hit in the provided logs. However, are you able to provide more details about your experience, as well as a log?

**Comment by greyson-signal** [comment]
@pgerber Are things still the same? Are you able to get the logs for A4? Your logs indicate that your device *thinks* that not all of you linked devices support sealed sender. The desktop logs you provided indicate that they reported themselves as sealed-sender-capable, but we don't have A4, so that's an unknown. 

The other thing you could try is unlinking your devices one-by-one and seeing if that helps. If you do that,  I'd follow these steps:

1. Unlink a device via the Signal Android settings.
2. Close the app (swipe it away in the app switcher).
3. Open the app again.
4. Try sending a message.
5. Grab a debug log.
6. Repeat until you can send sealed sender messages.

Then post what happened along with your debug logs. 

Note: You can safely unlink+relink a desktop instance without losing your messages (as long as you're re-linking a device with the same phone number, which in this case you would be).

**Comment by pgerber** [comment]
Yes, things are still the same.

Here is [the log from A4](https://debuglogs.org/1922c6bcdb54fb5f31550dff2e14f59b376c27611529a8273253432609ed094b)

I'll try to find some time for relinking all Desktop installations but that can take a while.

**Comment by pgerber** [comment]
I found some time to trying to relink sooner than expected.

So, this is what I did.

1. I unliked A2 and sealed sender appeared to work afterwards. So, I relinked it again. Only then I realized that sealed sender only worked for incoming messages. Both, outgoing messages while A2 was unliked and after relink were not sealed.

    Logs:
    [A1](https://debuglogs.org/74286233c30fd6fa2acd14c658629afa18e252ecb8d6471760c93e842892816f)
    [A2](https://debuglogs.org/9ed2f1f25422c32343e80187d74252a899e79f87385c08fd6259468ced393841)


2. Next I unliked A3. Afterwards sealed sender worked in both directions. However, ever since I relinked this device again A2 no longer synces it's message to the sibling devices. The recipient gets the message but the other Desktop and Android don't.

    [A1](https://debuglogs.org/74286233c30fd6fa2acd14c658629afa18e252ecb8d6471760c93e842892816f)
    [A3](https://debuglogs.org/74286233c30fd6fa2acd14c658629afa18e252ecb8d6471760c93e842892816f)
    [A2 (outgoing messages that aren't synced)](https://debuglogs.org/3ca4eea4efe6f326e2f6db057f440c4c9852edc9851b1dae3da2f1bdb374bd7e)

FYI, all desktop installations have been upgraded to v1.18.0-beta.6 in the meantime.

**Comment by greyson-signal** [comment]
So I see a single unlinking of device ID 5  in your Android (A1) log. After it's unlinked, you gain the ability to sent sealed sender messages again.

```java
2018-11-08 21:04:03.188 GMT+01:00 W PushServiceSocket: Opening URL: https://textsecure-service.whispersystems.org/v1/devices/5
...
2018-11-08 21:04:03.894 GMT+01:00 I RefreshUnidentifiedDeliveryAbilityJob: Set UD status to: true
```

It looks like you don't have a complete Android log though, since I only see that one unlink. Can you tell me if you submitted the Android log before or after A3 was unlinked?

**Comment by pgerber** [comment]
Oh, I linked the same log twice. I updated the link above. The first A1 log is after unlinked A2 and the other one after unlinking A3.

**Comment by greyson-signal** [comment]
@pgerber Ah thanks :) That confirms the previous unlink I saw was for A2, and now I see A3. Looks like your Android device has the ability to send sealed messages after the unlink of A3 as well.

The behavior you're describing, particularly what happens after step 2, is odd. At this point, it could be an issue where the server has bad state around whether all of your linked devices are sealed-sender-capable. If you're comfortable, could you send a message from the problem device to one of my test numbers? +1 610-549-1426. No worries if you don't want to. Thanks!

**Comment by greyson-signal** [comment]
@pgerber Still doing some investigation on why the original linking process resulted in signal thinking some one of your devices isn't sealed-sender-capable, but there should be a fix for the second issue you experienced (A2 not syncing outgoing messages) in a future desktop release. In the meantime, relinking A2 again might fix the problem.

**Comment by pgerber** [comment]
I believe this has been fixed long since in Signal Desktop. @jpds, @minnmann, let me know if this is still an issue for you.
